Building isolated workspace app fails when patching

<!--
  If this issue affects many people in a company/big team, create a post for your company in the following discussion:
  https://github.com/pnpm/pnpm/discussions/3787
  and link the issue in your post.

  This will help us prioritize issues that affect more people.
-->

### pnpm version: 7.9.0

### Expected behavior:
Installing dependencies successfully.

### Actual behavior:
I have a workspace with many apps, and some of those apps have patched dependencies. When creating a docker image to build `app A`, which doesn't need any patch, I only copy inside the container the `app A` folder and the shared libs it depends on. When pnpm is installing the dependencies, it fails with this error:
> ERR_PNPM_PATCH_NOT_APPLIED: The following patches were not applied: ...

I know that the issue is because `app A` doesn't need any patch, I tried moving the patches and the `patchedDependencies` property to each app package.json but patching didn't work. Am I doing something wrong?

Right now the workaround is to manually remove the `patchedDependencies` from the root `package.json` everytime I'm going to build one of the apps.

### Additional information:

 - `node -v` prints: v16.13.2
 - Windows, macOS, or Linux?: macOS

--- a/.changeset/angry-hounds-bow.md
+++ b/.changeset/angry-hounds-bow.md
@@ -5,4 +5,18 @@
 "@pnpm/types": minor
 ---
 
-add flag to make non-applied patches a warning instead of a fatal error
+A new setting supported in the pnpm section of the `package.json` file: `allowNonAppliedPatches`. When it is set to `true`, non-applied patches will not cause an error, just a warning will be printed. For example:
+
+```json
+{
+  "name": "foo",
+  "version": "1.0.0",
+  "pnpm": {
+    "patchedDependencies": {
+      "express@4.18.1": "patches/express@4.18.1.patch"
+    },
+    "allowNonAppliedPatches": true
+  }
+}
+```
+

--- a/.changeset/angry-hounds-bow.md
+++ b/.changeset/angry-hounds-bow.md
@@ -1,0 +1,8 @@
+---
+"@pnpm/core": minor
+"@pnpm/plugin-commands-installation": minor
+"@pnpm/resolve-dependencies": minor
+"@pnpm/types": minor
+---
+
+add flag to make non-applied patches a warning instead of a fatal error

--- a/.changeset/angry-hounds-bow.md
+++ b/.changeset/angry-hounds-bow.md
@@ -3,6 +3,7 @@
 "@pnpm/plugin-commands-installation": minor
 "@pnpm/resolve-dependencies": minor
 "@pnpm/types": minor
+"pnpm": minor
 ---
 
 A new setting supported in the pnpm section of the `package.json` file: `allowNonAppliedPatches`. When it is set to `true`, non-applied patches will not cause an error, just a warning will be printed. For example:

--- a/packages/core/src/getPeerDependencyIssues.ts
+++ b/packages/core/src/getPeerDependencyIssues.ts
@@ -51,6 +51,7 @@ export async function getPeerDependencyIssues (
     {
       currentLockfile: ctx.currentLockfile,
       allowedDeprecatedVersions: {},
+      allowNonAppliedPatches: false,
       defaultUpdateDepth: -1,
       dryRun: true,
       engineStrict: false,

--- a/packages/core/src/install/extendInstallOptions.ts
+++ b/packages/core/src/install/extendInstallOptions.ts
@@ -98,6 +98,7 @@ export interface StrictInstallOptions {
   modulesCacheMaxAge: number
   peerDependencyRules: PeerDependencyRules
   allowedDeprecatedVersions: AllowedDeprecatedVersions
+  allowNonAppliedPatches: boolean
   preferSymlinkedExecutables: boolean
   resolutionMode: 'highest' | 'time-based'
 
@@ -124,6 +125,7 @@ const defaults = async (opts: InstallOptions) => {
   }
   return {
     allowedDeprecatedVersions: {},
+    allowNonAppliedPatches: false,
     autoInstallPeers: false,
     childConcurrency: 5,
     depth: 0,

--- a/packages/core/src/install/index.ts
+++ b/packages/core/src/install/index.ts
@@ -819,6 +819,7 @@ const _installInContext: InstallFunction = async (projects, ctx, opts) => {
     {
       allowBuild: createAllowBuildFunction(opts),
       allowedDeprecatedVersions: opts.allowedDeprecatedVersions,
+      allowNonAppliedPatches: opts.allowNonAppliedPatches,
       autoInstallPeers: opts.autoInstallPeers,
       currentLockfile: ctx.currentLockfile,
       defaultUpdateDepth: (opts.update || (opts.updateMatching != null)) ? opts.depth : -1,

--- a/packages/core/test/install/patch.ts
+++ b/packages/core/test/install/patch.ts
@@ -96,6 +96,28 @@ test('patch package', async () => {
   expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).not.toContain('// patched')
 })
 
+test('patch package reports warning if not all patches are applied and allowNonAppliedPatches is set', async () => {
+  prepareEmpty()
+  const patchPath = path.join(f.find('patch-pkg'), 'is-positive@1.0.0.patch')
+
+  const patchedDependencies = {
+    'is-positive@1.0.0': path.relative(process.cwd(), patchPath),
+    'is-negative@1.0.0': path.relative(process.cwd(), patchPath),
+  }
+  const opts = await testDefaults({
+    fastUnpack: false,
+    sideEffectsCacheRead: true,
+    sideEffectsCacheWrite: true,
+    patchedDependencies,
+    allowNonAppliedPatches: true,
+  }, {}, {}, { packageImportMethod: 'hardlink' })
+  await install({
+    dependencies: {
+      'is-positive': '1.0.0',
+    },
+  }, opts)
+})
+
 test('patch package throws an exception if not all patches are applied', async () => {
   prepareEmpty()
   const patchPath = path.join(f.find('patch-pkg'), 'is-positive@1.0.0.patch')

--- a/packages/core/test/install/patch.ts
+++ b/packages/core/test/install/patch.ts
@@ -98,6 +98,7 @@ test('patch package', async () => {
 
 test('patch package reports warning if not all patches are applied and allowNonAppliedPatches is set', async () => {
   prepareEmpty()
+  const reporter = jest.fn()
   const patchPath = path.join(f.find('patch-pkg'), 'is-positive@1.0.0.patch')
 
   const patchedDependencies = {
@@ -110,12 +111,19 @@ test('patch package reports warning if not all patches are applied and allowNonA
     sideEffectsCacheWrite: true,
     patchedDependencies,
     allowNonAppliedPatches: true,
+    reporter,
   }, {}, {}, { packageImportMethod: 'hardlink' })
   await install({
     dependencies: {
       'is-positive': '1.0.0',
     },
   }, opts)
+  expect(reporter).toBeCalledWith(
+    expect.objectContaining({
+      level: 'warn',
+      message: 'The following patches were not applied: is-negative@1.0.0',
+    })
+  )
 })
 
 test('patch package throws an exception if not all patches are applied', async () => {

--- a/packages/plugin-commands-installation/src/getOptionsFromRootManifest.ts
+++ b/packages/plugin-commands-installation/src/getOptionsFromRootManifest.ts
@@ -7,6 +7,7 @@ import {
 
 export default function getOptionsFromRootManifest (manifest: ProjectManifest): {
   allowedDeprecatedVersions?: AllowedDeprecatedVersions
+  allowNonAppliedPatches?: boolean
   overrides?: Record<string, string>
   neverBuiltDependencies?: string[]
   onlyBuiltDependencies?: string[]
@@ -23,9 +24,11 @@ export default function getOptionsFromRootManifest (manifest: ProjectManifest): 
   const packageExtensions = manifest.pnpm?.packageExtensions
   const peerDependencyRules = manifest.pnpm?.peerDependencyRules
   const allowedDeprecatedVersions = manifest.pnpm?.allowedDeprecatedVersions
+  const allowNonAppliedPatches = manifest.pnpm?.allowNonAppliedPatches
   const patchedDependencies = manifest.pnpm?.patchedDependencies
   const settings = {
     allowedDeprecatedVersions,
+    allowNonAppliedPatches,
     overrides,
     neverBuiltDependencies,
     packageExtensions,

--- a/packages/resolve-dependencies/src/index.ts
+++ b/packages/resolve-dependencies/src/index.ts
@@ -4,7 +4,7 @@ import PnpmError from '@pnpm/error'
 import {
   packageManifestLogger,
 } from '@pnpm/core-loggers'
-import logger from '@pnpm/logger'
+import globalWarn from '@pnpm/logger'
 import {
   Lockfile,
   ProjectSnapshot,
@@ -271,15 +271,11 @@ export default async function (
 function verifyPatches (patchedDependencies: string[], appliedPatches: Set<string>, allowNonAppliedPatches: boolean = false) {
   const nonAppliedPatches: string[] = patchedDependencies.filter((patchKey) => !appliedPatches.has(patchKey))
   if (nonAppliedPatches.length) {
-    const code = 'PATCH_NOT_APPLIED'
     const message = `The following patches were not applied: ${nonAppliedPatches.join(', ')}`
     if (allowNonAppliedPatches) {
-      logger.warn({
-        prefix: code,
-        message,
-      })
+      globalWarn(message)
     } else {
-      throw new PnpmError(code, message, {
+      throw new PnpmError('PATCH_NOT_APPLIED', message, {
         hint: 'Either remove them from "patchedDependencies" or update them to match packages in your dependencies.',
       })
     }

--- a/packages/resolve-dependencies/src/index.ts
+++ b/packages/resolve-dependencies/src/index.ts
@@ -4,6 +4,7 @@ import PnpmError from '@pnpm/error'
 import {
   packageManifestLogger,
 } from '@pnpm/core-loggers'
+import logger from '@pnpm/logger'
 import {
   Lockfile,
   ProjectSnapshot,
@@ -83,6 +84,7 @@ export default async function (
     preserveWorkspaceProtocol: boolean
     saveWorkspaceProtocol: 'rolling' | boolean
     lockfileIncludeTarballUrl?: boolean
+    allowNonAppliedPatches?: boolean
   }
 ) {
   const _toResolveImporter = toResolveImporter.bind(null, {
@@ -110,7 +112,7 @@ export default async function (
     (opts.forceFullResolution || !opts.wantedLockfile.packages?.length) &&
     Object.keys(opts.wantedLockfile.importers).length === importers.length
   ) {
-    verifyPatches(Object.keys(opts.patchedDependencies), appliedPatches)
+    verifyPatches(Object.keys(opts.patchedDependencies), appliedPatches, opts.allowNonAppliedPatches)
   }
 
   const linkedDependenciesByProjectId: Record<string, LinkedDependency[]> = {}
@@ -266,12 +268,21 @@ export default async function (
   }
 }
 
-function verifyPatches (patchedDependencies: string[], appliedPatches: Set<string>) {
+function verifyPatches (patchedDependencies: string[], appliedPatches: Set<string>, allowNonAppliedPatches: boolean = false) {
   const nonAppliedPatches: string[] = patchedDependencies.filter((patchKey) => !appliedPatches.has(patchKey))
   if (nonAppliedPatches.length) {
-    throw new PnpmError('PATCH_NOT_APPLIED', `The following patches were not applied: ${nonAppliedPatches.join(', ')}`, {
-      hint: 'Either remove them from "patchedDependencies" or update them to match packages in your dependencies.',
-    })
+    const code = 'PATCH_NOT_APPLIED'
+    const message = `The following patches were not applied: ${nonAppliedPatches.join(', ')}`
+    if (allowNonAppliedPatches) {
+      logger.warn({
+        prefix: code,
+        message,
+      })
+    } else {
+      throw new PnpmError(code, message, {
+        hint: 'Either remove them from "patchedDependencies" or update them to match packages in your dependencies.',
+      })
+    }
   }
 }
 

--- a/packages/resolve-dependencies/src/index.ts
+++ b/packages/resolve-dependencies/src/index.ts
@@ -4,7 +4,7 @@ import PnpmError from '@pnpm/error'
 import {
   packageManifestLogger,
 } from '@pnpm/core-loggers'
-import globalWarn from '@pnpm/logger'
+import { globalWarn } from '@pnpm/logger'
 import {
   Lockfile,
   ProjectSnapshot,
@@ -112,7 +112,11 @@ export default async function (
     (opts.forceFullResolution || !opts.wantedLockfile.packages?.length) &&
     Object.keys(opts.wantedLockfile.importers).length === importers.length
   ) {
-    verifyPatches(Object.keys(opts.patchedDependencies), appliedPatches, opts.allowNonAppliedPatches)
+    verifyPatches({
+      patchedDependencies: Object.keys(opts.patchedDependencies),
+      appliedPatches,
+      allowNonAppliedPatches: opts.allowNonAppliedPatches,
+    })
   }
 
   const linkedDependenciesByProjectId: Record<string, LinkedDependency[]> = {}
@@ -268,18 +272,27 @@ export default async function (
   }
 }
 
-function verifyPatches (patchedDependencies: string[], appliedPatches: Set<string>, allowNonAppliedPatches: boolean = false) {
-  const nonAppliedPatches: string[] = patchedDependencies.filter((patchKey) => !appliedPatches.has(patchKey))
-  if (nonAppliedPatches.length) {
-    const message = `The following patches were not applied: ${nonAppliedPatches.join(', ')}`
-    if (allowNonAppliedPatches) {
-      globalWarn(message)
-    } else {
-      throw new PnpmError('PATCH_NOT_APPLIED', message, {
-        hint: 'Either remove them from "patchedDependencies" or update them to match packages in your dependencies.',
-      })
-    }
+function verifyPatches (
+  {
+    patchedDependencies,
+    appliedPatches,
+    allowNonAppliedPatches,
+  }: {
+    patchedDependencies: string[]
+    appliedPatches: Set<string>
+    allowNonAppliedPatches: boolean
   }
+): void {
+  const nonAppliedPatches: string[] = patchedDependencies.filter((patchKey) => !appliedPatches.has(patchKey))
+  if (!nonAppliedPatches.length) return
+  const message = `The following patches were not applied: ${nonAppliedPatches.join(', ')}`
+  if (allowNonAppliedPatches) {
+    globalWarn(message)
+    return
+  }
+  throw new PnpmError('PATCH_NOT_APPLIED', message, {
+    hint: 'Either remove them from "patchedDependencies" or update them to match packages in your dependencies.',
+  })
 }
 
 async function finishLockfileUpdates (

--- a/packages/resolve-dependencies/src/resolveDependencyTree.ts
+++ b/packages/resolve-dependencies/src/resolveDependencyTree.ts
@@ -62,6 +62,7 @@ export interface ResolveDependenciesOptions {
   autoInstallPeers?: boolean
   allowBuild?: (pkgName: string) => boolean
   allowedDeprecatedVersions: AllowedDeprecatedVersions
+  allowNonAppliedPatches: boolean
   currentLockfile: Lockfile
   dryRun: boolean
   engineStrict: boolean

--- a/packages/types/src/package.ts
+++ b/packages/types/src/package.ts
@@ -129,6 +129,7 @@ export type ProjectManifest = BaseManifest & {
     packageExtensions?: Record<string, PackageExtension>
     peerDependencyRules?: PeerDependencyRules
     allowedDeprecatedVersions?: AllowedDeprecatedVersions
+    allowNonAppliedPatches?: boolean
     patchedDependencies?: Record<string, string>
   }
   private?: boolean

